org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/SaveActionHandler.java

```diff
@@ -48,17 +60,55 @@ public List<TextEdit> willSaveWaitUntil(WillSaveTextDocumentParams params, IProg
 
 		String documentUri = params.getTextDocument().getUri();
 
-		if (preferenceManager.getPreferences().isJavaSaveActionsOrganizeImportsEnabled()) {
+		Preferences preferences = preferenceManager.getPreferences();
+		IEclipsePreferences jdtUiPreferences = getJdtUiProjectPreferences(documentUri);
+		if (preferences.isJavaSaveActionsOrganizeImportsEnabled() ||
+			(jdtUiPreferences != null && jdtUiPreferences.getBoolean("sp_" + CleanUpConstants.ORGANIZE_IMPORTS, false))) {
 			edit.addAll(handleSaveActionOrganizeImports(documentUri, monitor));
 		}
 
-		Preferences preferences = preferenceManager.getPreferences();
-		List<TextEdit> cleanUpEdits = cleanUpRegistry.getEditsForAllActiveCleanUps(params.getTextDocument(), preferences.getCleanUpActionsOnSave(), monitor);
+		LinkedHashSet<String> cleanUpIds = new LinkedHashSet<>();
+		cleanUpIds.addAll(preferences.getCleanUpActionsOnSave());
+		cleanUpIds.addAll(getCleanupIdsForProjectPreferences(jdtUiPreferences));
```

Comment:
I would maybe call it `getCleanupsFromJDTUIPreferences(..)` ? Or at least something to make it clear this probably does nothing when run outside of the Eclipse platform.

Reply:
OK, I'll change it.

I also wonder whether it couldn't be better to just change the current ISimpleCleanUp so it can return multiple ids directly, not making a different between JDT-LS and JDT-UI one. So instead of a new method, it would look like

```java
public class AddOverrideAnnotationCleanUp implements ISimpleCleanUp {

	@Override
	public Collection<String> getIdentifiers() {
		return List.of("addOverride", CleanUpConstants.ADD_MISSING_ANNOTATIONS_OVERRIDE);
	}
```
What do you think about it?

Reply:
I like this approach a lot more. However it complicates the check to determine whether the ID exists a bit. Basically we'd be doing any `anyMatch` style check against the list of enabled cleanups, for every clean up. I guess that wouldn't be so bad..

Reply:
This is not much more complicated: https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/2870/files#diff-17111006f1acef68e115f79c8448d4dfe6d1821d1957a456277e6988ee82428bR62-R63